pkg/cli/dump.go

```diff
@@ -260,30 +285,136 @@ type tableMetadata struct {
 	basicMetadata
 
 	columnNames string
-	columnTypes map[string]*types.T
+	columnTypes map[string]tree.ResolvableTypeReference
 }
 
-// getDumpMetadata retrieves the table information for the specified table(s).
-// It also retrieves the cluster timestamp at which the metadata was
-// retrieved.
-func getDumpMetadata(
-	conn *sqlConn, dbName string, tableNames []string, asOf string,
-) (mds []basicMetadata, err error) {
+// dumpTypeContext acts as a collection of user defined types to resolve
+// references to user defined types in the dump process.
+type dumpTypeContext struct {
+	typMap           map[types.UserDefinedTypeName]*types.T
+	createStatements []string
+}
+
+// ResolveType implements the tree.TypeReferenceResolver interface.
+func (d *dumpTypeContext) ResolveType(
+	_ context.Context, name *tree.UnresolvedObjectName,
+) (*types.T, error) {
+	key := types.UserDefinedTypeName{
+		Name:    name.Object(),
+		Schema:  name.Schema(),
+		Catalog: name.Catalog(),
+	}
+	typ, ok := d.typMap[key]
+	if !ok {
+		return nil, errors.Newf("type %s not found", name.String())
+	}
+	return typ, nil
+}
+
+// ResolveTypeByID implements the tree.TypeReferenceResolver interface.
+func (d *dumpTypeContext) ResolveTypeByID(context.Context, uint32) (*types.T, error) {
+	return nil, errors.AssertionFailedf("cannot resolve types in dump by ID")
+}
+
+// collectUserDefinedTypes constructs a dumpTypeContext consisting of all user
+// defined types in the requested database.
+func collectUserDefinedTypes(conn *sqlConn, dbName string, ts string) (*dumpTypeContext, error) {
+	query := `
+SELECT
+	descriptor_id, create_statement
+FROM
+	"".crdb_internal.create_type_statements
+AS OF SYSTEM TIME %s
+WHERE 
+	database_name = $1
+`
+	rows, err := conn.Query(fmt.Sprintf(query, lex.EscapeSQLString(ts)), []driver.Value{dbName})
+	if err != nil {
+		return nil, err
+	}
+	vals := make([]driver.Value, 2)
+	var createStatements []string
+	typContext := &dumpTypeContext{
+		typMap: make(map[types.UserDefinedTypeName]*types.T),
+	}
+	for {
+		if err := rows.Next(vals); err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+		idI := vals[0]
+		id, ok := idI.(int64)
+		if !ok {
+			return nil, errors.AssertionFailedf("unexpected value %T", idI)
+		}
+
+		createStatementI := vals[1]
+		createStatement, ok := createStatementI.(string)
+		if !ok {
+			return nil, errors.AssertionFailedf("unexpected value %T", createStatementI)
+		}
+		createStatements = append(createStatements, createStatement)
+
+		// Parse the create statement into the corresponding AST node. This is done
+		// so that information like the qualified name and enum members can be
+		// parsed out of the definition.
+		stmt, err := parser.ParseOne(createStatement)
```

Comment:
ok here you're getting a step too far. `dump` should have no business interpreting the strings coming from the server, given that it's likely to run a version behind.

Instead I recommend you operate as follows:

1. make the server able to produce the information you need, via a new `SHOW` statement or `crdb_internal` vtable
2. use it here without interpretaiton

Reply:
heh, yeah I felt that I was taking it a step too far here. I'll extend the internal table I'm referencing to return the necessary information.